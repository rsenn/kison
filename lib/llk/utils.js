// @ts-check

const data = require('../data');

let { cachedStateMatchMap } = data;

function setStateCacheMatch(state, tokensLength) {
  return
  let holder = cachedStateMatchMap.get(state);
  if (!holder) {
    holder = new Set();
    cachedStateMatchMap.set(state, holder);
  }
  holder.add(tokensLength);
}

function getStateCacheMatch(state, tokensLength) {
  return;
  // consumed token length may be different!
  const holder = cachedStateMatchMap.get(state);
  return holder?.has(tokensLength);
}

function clearStateMatchCache() {
  return
  cachedStateMatchMap = new Map();
}

module.exports = {
  setStateCacheMatch,
  getStateCacheMatch,
  clearStateMatchCache,
};