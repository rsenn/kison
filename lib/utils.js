/**
 * @ignore
 * utils for kison.
 * @author yiminghe@gmail.com
 */

var util = require("modulex-util");

var doubleReg = /"/g;
var single = /'/g;

const gened = {
  mix(to, from) {
    for (var f in from) {
      to[f] = from[f];
    }
  },

  isArray(obj) {
    return "[object Array]" === Object.prototype.toString.call(obj);
  },

  inArray(item, arr) {
    for (var i = 0, l = arr.length; i < l; i++) {
      if (arr[i] === item) {
        return true;
      }
    }
    return false;
  },

  each(object, fn, context) {
    if (object) {
      var key,
        val,
        length,
        i = 0;

      context = context || null;

      if (!isArray(object)) {
        for (key in object) {
          // can not use hasOwnProperty
          if (fn.call(context, object[key], key, object) === false) {
            break;
          }
        }
      } else {
        length = object.length;
        for (val = object[0]; i < length; val = object[++i]) {
          if (fn.call(context, val, i, object) === false) {
            break;
          }
        }
      }
    }
  }
};

var { isArray } = gened;

const utils = (module.exports = Object.assign({}, util, gened, {
  eachRhs(rhs, fn) {
    utils.each(rhs, (r, index) => {
      if (typeof r !== "string") {
        return;
      }
      fn(r, index);
    });
  },

  assertSymbolString(symbol) {
    if (typeof symbol !== "string") {
      throw new Error("invalid symbol".symbol);
    }
  },

  filterRhs(rhs) {
    return rhs.filter(r => typeof r === "string");
  },

  toFunctionString(fn, name) {
    const str = fn.toString();
    if (str.match(/^function\s+\(/)) {
      return str.replace(/^function\s+\(/, `function ${name || fn.name}(`);
    }
    return "function " + str;
  },

  escapeString(str, quote) {
    var regexp = single;
    if (quote === '"') {
      regexp = doubleReg;
    } else {
      quote = "'";
    }
    return str
      .replace(/\\/g, "\\\\")
      .replace(/\r/g, "\\r")
      .replace(/\n/g, "\\n")
      .replace(/\t/g, "\\t")
      .replace(regexp, "\\" + quote);
  },

  serializeObject(obj, excludeReg) {
    var r;

    if (
      excludeReg &&
      typeof excludeReg === "function" &&
      (r = excludeReg(obj)) === false
    ) {
      return false;
    }

    if (r !== undefined) {
      obj = r;
    }

    var ret = [];

    if (typeof obj === "string") {
      return "'" + utils.escapeString(obj) + "'";
    } else if (typeof obj === "number") {
      return obj + "";
    } else if (util.isRegExp(obj)) {
      return (
        "/" +
        obj.source +
        "/" +
        (obj.global ? "g" : "") +
        (obj.ignoreCase ? "i" : "") +
        (obj.multiline ? "m" : "")
      );
    } else if (util.isArray(obj)) {
      ret.push("[");
      var sub = [];
      util.each(obj, function(v) {
        var t = utils.serializeObject(v, excludeReg);
        if (t !== false) {
          sub.push(t);
        }
      });
      ret.push(sub.join(", "));
      ret.push("]");
      return ret.join("");
    } else if (typeof obj === "object") {
      ret = [];
      ret[0] = "{";
      var start = 1;
      for (var i in obj) {
        var v = obj[i];
        if (excludeReg && util.isRegExp(excludeReg) && i.match(excludeReg)) {
          continue;
        }
        var t = utils.serializeObject(v, excludeReg);
        if (t === false) {
          continue;
        }
        /*jshint quotmark:false*/
        var key = "'" + utils.escapeString(i) + "'";
        ret.push((start ? "" : ",") + key + ": " + t);
        start = 0;
      }
      ret.push("}");
      return ret.join("\n");
    } else if (typeof obj === "function") {
      return (obj + "").replace(/^[\w-]+/, "function");
    } else {
      return obj + "";
    }
  }
}));
