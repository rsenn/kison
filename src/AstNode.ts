import type { ParseError } from './types';
import data from './data';

interface Params {
  parent?: AstNode;
}

interface TokenParams extends Params {
  token?: string;
  text?: string;
  t?: string;
}

interface ErrorTokenParams extends TokenParams {
  error?: ParseError;
}

interface SymbolParams extends Params {
  id: number;
  children?: AstNode[];
  symbol?: string;
  label?: string;
  internalRuleIndex?: number;
}

export type AstNode = AstSymbolNode | AstTokenNode;

export class BaseAstNode {
  parent?: AstSymbolNode;
  start: number = 0;
  end: number = 0;
  firstLine: number = 0;
  lastLine: number = 0;
  firstColumn: number = 0;
  lastColumn: number = 0;

  toJSON() {
    const ret: Record<string, any> = {};
    for (const k of Object.keys(this)) {
      if (k !== 'parent' && k !== 't') {
        const v = (this as any)[k];
        if (v !== undefined) {
          ret[k] = v;
        }
      }
    }
    return ret;
  }
}

export class AstTokenNode extends BaseAstNode {
  token: string = '';
  t: string = '';
  recovery?: string;
  type: 'token' = 'token';
  constructor(params: TokenParams) {
    super();
    Object.assign(this, params);
  }
}

export class AstErrorNode extends AstTokenNode {
  error?: ParseError;
  constructor(ErrorTokenParams: ErrorTokenParams) {
    super(ErrorTokenParams);
    Object.assign(ErrorTokenParams);
  }
}

const { productionRuleIndexMap } = data;

export class AstSymbolNode extends BaseAstNode {
  symbol: string = '';
  label?: string;
  type: 'symbol' = 'symbol';
  children: AstNode[] = [];
  ruleIndex: number = -1;
  internalRuleIndex: number = -1;
  constructor(params: SymbolParams) {
    super();
    Object.assign(this, params);
    if (params.children) {
      this.setChildren(params.children);
    }
    if (params.internalRuleIndex !== undefined) {
      this.ruleIndex = productionRuleIndexMap[this.internalRuleIndex];
    }
  }

  addChild(c: AstNode) {
    this.addChildren([c]);
  }

  addChildren(cs: AstNode[]) {
    this.children = this.children || [];
    this.children.push(...cs);
    this.setChildren(this.children);
  }

  refreshChildren() {
    this.setChildren(this.children);
  }

  setChildren(cs: AstNode[]) {
    if (!cs.length) {
      this.children = [];
      return;
    }
    const first = cs[0];
    const last = cs[cs.length - 1];
    this.start = first.start;
    this.end = last.end;
    this.firstLine = first.firstLine;
    this.lastLine = last.lastLine;
    this.firstColumn = first.firstColumn;
    this.lastColumn = last.lastColumn;
    this.children = cs;
    for (const c of cs) {
      c.parent = this;
    }
  }
}
