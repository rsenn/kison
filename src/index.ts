// @ts-ignore
import Utils from 'modulex-util';
import LLGrammar from './ll/LLGrammar';
import LLKGrammar from './llk/LLKGrammar';
import LALRGrammar from './lalr/LALRGrammar';
import fs from 'fs';
import type { WriteFileOptions } from 'fs';
import path from 'path';
import program from 'commander';
import pkg from '../package.json';
import { build } from 'esbuild';
import { options as KisonOptions } from './options';

var encoding: WriteFileOptions = 'utf-8';

const ConsMap: Record<string, any> = {
  ll: LLGrammar,
  lalr: LALRGrammar,
  llk: LLKGrammar,
};
const placehoder = '__KISON___GENERATED__CODE__';

program
  .option('-g, --grammar <grammar>', 'Set kison grammar file')
  .option('-o, --output [file]', 'output file path')
  .option('--library [library]', 'library name')
  .option(
    '--astNodeUserDataTypes [astNodeUserDataTypes]',
    'astNodeUserDataTypes file',
  )
  .option('--bnf [bnf]', 'bnf file')
  .option('--declaration [declaration]', 'declaration')
  .option('--declarationDir [declarationDir]', 'declarationDir')
  .option('-m, --mode [mode]', 'lalr or ll')
  .option('-b, --babel [babel]', 'use babel')
  .option('-v, --verbose [verbose]', 'verbose')
  .option('--es [es]', 'generate es module')
  // defaults bool true
  .option('--no-compressSymbol', 'Set compress symbol')
  .option('--no-compressLexerState', 'Set compress lexer state')
  .parse(process.argv);

var options = program.options;

var grammar = path.resolve(program.grammar);

options.forEach(function (o: any) {
  var name = o.name();
  if (o.required && !(name in program)) {
    program.optionMissingArgument(o);
  }
});

var kisonCfg = {
  compressLexerState: program.compressLexerState,
  compressSymbol: program.compressSymbol,
};

var outFile = program.output
  ? program.output
  : path.basename(grammar, '-grammar.js');

var grammarBaseName = program.library || '$parser';

const mode = program.mode || 'lalr';

if (!ConsMap[mode]) {
  throw new Error('Unsupported mode: ' + mode);
}

var modulePath = program.output
  ? path.resolve(outFile)
  : path.resolve(grammar, '../' + outFile + '.js');

var codeTemplate = `
/*
Generated By kison v${pkg.version}

Generate time: ${new Date() + ''}
*/
 var ${grammarBaseName} = ${placehoder};
`;

if (program.es) {
  codeTemplate += `
   export default ${grammarBaseName};
`;
} else {
  codeTemplate += `
 if(typeof module !== 'undefined') {
   module.exports = ${grammarBaseName};
 }
`;
}

function myBabel(str: string) {
  try {
    require('@babel/core');
    require('@babel/preset-env');
  } catch (e) {
    console.log(
      'Info: you need to install @babel/core@7.x and @babel/preset-env@7.x manually so kison will use it to trans generated code.\n',
    );
    return str;
  }
  const pkg = require('@babel/core/package.json');
  if (!pkg.version.startsWith('7.')) {
    return str;
  }
  var babel = require('@babel/core');
  let transformRet;
  try {
    transformRet = babel.transformSync(str, {
      presets: [
        [
          require.resolve('@babel/preset-env'),
          {
            loose: true,
          },
        ],
      ],
    }); // => { code, map, ast }
  } catch (e) {
    console.log(e);
    return str;
  }
  return transformRet.code;
}

function myJsBeautify(str: string) {
  let prettier;
  try {
    prettier = require('prettier');
  } catch (e) {
    console.log(
      'Info: you can install prettier manually so kison will use it to format generated code.\n',
    );
    return str;
  }
  try {
    return prettier.format(str, {
      parser: 'babel',
    });
  } catch (e) {
    console.log(e);
  }
  return str;
}

async function bundleConfigFile(fileName: string): Promise<string> {
  const result = await build({
    absWorkingDir: process.cwd(),
    entryPoints: [fileName],
    outfile: 'out.js',
    write: false,
    platform: 'node',
    bundle: true,
    format: 'cjs',
    sourcemap: 'inline',
    metafile: true,
    plugins: [
      {
        name: 'externalize-deps',
        setup(build) {
          build.onResolve({ filter: /.*/ }, (args) => {
            const id = args.path;
            if (id[0] !== '.' && !path.isAbsolute(id)) {
              return {
                external: true,
              };
            }
          });
        },
      },
      {
        name: 'replace-import-meta',
        setup(build) {
          build.onLoad({ filter: /\.[jt]s$/ }, async (args) => {
            const contents = await fs.promises.readFile(args.path, 'utf8');
            return {
              loader: args.path.endsWith('.ts') ? 'ts' : 'js',
              contents: contents
                .replace(
                  /\bimport\.meta\.url\b/g,
                  JSON.stringify(`file://${args.path}`),
                )
                .replace(
                  /\b__dirname\b/g,
                  JSON.stringify(path.dirname(args.path)),
                )
                .replace(/\b__filename\b/g, JSON.stringify(args.path)),
            };
          });
        },
      },
    ],
  });
  const { text } = result.outputFiles[0];
  return text;
}

async function genParser() {
  var grammarObj;
  if (grammar.endsWith('.js')) {
    grammarObj = require(grammar);
  } else {
    const code = await bundleConfigFile(grammar);
    const newFile = grammar.replace(/\.\w+$/, '.js');
    fs.writeFileSync(newFile, code);
    grammarObj = require(newFile);
    grammarObj = grammarObj.default || grammarObj;
    fs.unlinkSync(newFile);
  }

  if (typeof grammarObj === 'function') {
    grammarObj = grammarObj(KisonOptions);
  }

  const Cons = ConsMap[mode];

  console.info('start generate grammar module: ' + modulePath + '\n');
  var start = Date.now();

  const instance = new Cons(grammarObj);

  if (program.declaration && instance.genDTs) {
    let output = program.declarationDir;
    output = typeof output === 'string' ? output : path.dirname(outFile);
    output = path.resolve(output);
    const file = path.join(
      output,
      path.basename(outFile, path.extname(outFile)) + '.d.ts',
    );
    const baseDts = fs.readFileSync(
      path.join(__dirname, './parser.d.ts'),
      'utf-8',
    );
    let astNodeUserDataTypesPath = program.astNodeUserDataTypes;
    let astNodeUserDataTypes;
    if (astNodeUserDataTypesPath) {
      astNodeUserDataTypesPath = path.resolve(
        astNodeUserDataTypesPath + '.d.ts',
      );
      astNodeUserDataTypes = fs.readFileSync(astNodeUserDataTypesPath, 'utf-8');
    }
    const dts = instance.genDTs(baseDts, astNodeUserDataTypes);
    if (dts) {
      fs.writeFileSync(file, dts);
    }
  }

  if (program.bnf && instance.toBNF) {
    const bnf = instance.toBNF();
    let output = program.bnf;
    if (typeof output === 'string') {
      output = path.resolve(output);
      fs.writeFileSync(output, bnf);
    } else {
      console.log('');
      console.log(bnf);
      console.log('');
    }
  }

  /*jshint evil:true*/

  var code = instance.genCode(kisonCfg);

  var moduleCode = codeTemplate.replace(placehoder, () => code);

  if (program.babel) {
    moduleCode = myBabel(moduleCode);
  }

  moduleCode = myJsBeautify(moduleCode);

  fs.writeFileSync(modulePath, moduleCode, encoding);

  console.info(
    'generate grammar module: ' +
      modulePath +
      ' at ' +
      new Date().toLocaleString(),
  );
  console.log();
  console.info('duration: ' + (Date.now() - start) + 'ms');
  if (program.verbose) {
    console.log();
    console.log();
    console.log();
    console.log('******************* table');
    console.log(instance.visualizeTable());
  }
}

genParser();
