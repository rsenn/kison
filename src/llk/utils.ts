// @ts-check

import data from '../data';

let { productionsBySymbol } = data;

function isSymbol(s: string) {
  return !!productionsBySymbol[s];
}

export { isSymbol };
