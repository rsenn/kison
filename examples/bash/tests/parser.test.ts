import { prettyJson } from '../../../__tests__/utils';
import { Ast_Command_Node, Ast_String_Node, parser } from '../src/';

describe('bash parser', () => {
  it('token correctly', () => {
    const code = `
    curl "https://api.test.com/" -u "some_username:some_password"
`.trim();
    const ret = parser.lex(code);
    expect(ret.tokens).toMatchSnapshot();
  });

  it('token new line', () => {
    const code = `
    curl "x\\
    z" \\
     -u "a:b"
`.trim();
    const ret = parser.lex(code);
    expect(ret.tokens).toMatchSnapshot();
  });

  it('parse correctly', () => {
    const code = `
    curl "https://api.test.com/" -u "some_username:some_password"
`.trim();
    const ret = parser.parse(code);

    expect(ret.error?.errorMessage).toMatchInlineSnapshot(`undefined`);

    expect(ret.ast).toMatchSnapshot();
  });

  it('parse new line', () => {
    const code = `
    curl "x\\
z" \\
     -u "a:b"
`.trim();
    const ret = parser.parse(code);

    expect(ret.error?.errorMessage).toMatchInlineSnapshot(`undefined`);

    expect(ret.ast).toMatchSnapshot();

    const c = (ret.ast.children[0] as Ast_Command_Node).children[1]
      .children[0] as Ast_String_Node;
    expect(c.children[1].text).toBe('x\\\nz');
  });

  it('parse multiple line', () => {
    const code = `
    curl 'www.taobao.com'
    curl 'www.baidu.com'
    `;

    const ret = parser.parse(code);

    expect(ret.error?.errorMessage).toMatchInlineSnapshot(`undefined`);

    expect(ret.ast).toMatchSnapshot();
  });
});
