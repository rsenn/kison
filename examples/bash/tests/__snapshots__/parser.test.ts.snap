// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bash parser parse correctly 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "end": 4,
              "firstColumn": 1,
              "firstLine": 1,
              "lastColumn": 5,
              "lastLine": 1,
              "start": 0,
              "text": "curl",
              "token": "WORD",
              "type": "token",
            },
          ],
          "end": 4,
          "firstColumn": 1,
          "firstLine": 1,
          "id": 4,
          "internalRuleIndex": 5,
          "lastColumn": 5,
          "lastLine": 1,
          "ruleIndex": 3,
          "start": 0,
          "symbol": "literal",
          "type": "symbol",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "end": 6,
                  "firstColumn": 6,
                  "firstLine": 1,
                  "lastColumn": 7,
                  "lastLine": 1,
                  "start": 5,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
                Object {
                  "end": 27,
                  "firstColumn": 7,
                  "firstLine": 1,
                  "lastColumn": 28,
                  "lastLine": 1,
                  "start": 6,
                  "text": "https://api.test.com/",
                  "token": "STRING_CONTENT",
                  "type": "token",
                },
                Object {
                  "end": 28,
                  "firstColumn": 28,
                  "firstLine": 1,
                  "lastColumn": 29,
                  "lastLine": 1,
                  "start": 27,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
              ],
              "end": 28,
              "firstColumn": 6,
              "firstLine": 1,
              "id": 6,
              "internalRuleIndex": 10,
              "lastColumn": 29,
              "lastLine": 1,
              "ruleIndex": 4,
              "start": 5,
              "symbol": "string",
              "type": "symbol",
            },
          ],
          "end": 28,
          "firstColumn": 6,
          "firstLine": 1,
          "id": 5,
          "internalRuleIndex": 6,
          "lastColumn": 29,
          "lastLine": 1,
          "ruleIndex": 3,
          "start": 5,
          "symbol": "literal",
          "type": "symbol",
        },
        Object {
          "children": Array [
            Object {
              "end": 31,
              "firstColumn": 30,
              "firstLine": 1,
              "lastColumn": 32,
              "lastLine": 1,
              "start": 29,
              "text": "-u",
              "token": "WORD",
              "type": "token",
            },
          ],
          "end": 31,
          "firstColumn": 30,
          "firstLine": 1,
          "id": 7,
          "internalRuleIndex": 5,
          "lastColumn": 32,
          "lastLine": 1,
          "ruleIndex": 3,
          "start": 29,
          "symbol": "literal",
          "type": "symbol",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "end": 33,
                  "firstColumn": 33,
                  "firstLine": 1,
                  "lastColumn": 34,
                  "lastLine": 1,
                  "start": 32,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
                Object {
                  "end": 60,
                  "firstColumn": 34,
                  "firstLine": 1,
                  "lastColumn": 61,
                  "lastLine": 1,
                  "start": 33,
                  "text": "some_username:some_password",
                  "token": "STRING_CONTENT",
                  "type": "token",
                },
                Object {
                  "end": 61,
                  "firstColumn": 61,
                  "firstLine": 1,
                  "lastColumn": 62,
                  "lastLine": 1,
                  "start": 60,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
              ],
              "end": 61,
              "firstColumn": 33,
              "firstLine": 1,
              "id": 9,
              "internalRuleIndex": 10,
              "lastColumn": 62,
              "lastLine": 1,
              "ruleIndex": 4,
              "start": 32,
              "symbol": "string",
              "type": "symbol",
            },
          ],
          "end": 61,
          "firstColumn": 33,
          "firstLine": 1,
          "id": 8,
          "internalRuleIndex": 6,
          "lastColumn": 62,
          "lastLine": 1,
          "ruleIndex": 3,
          "start": 32,
          "symbol": "literal",
          "type": "symbol",
        },
      ],
      "end": 61,
      "firstColumn": 1,
      "firstLine": 1,
      "id": 3,
      "internalRuleIndex": 4,
      "lastColumn": 62,
      "lastLine": 1,
      "ruleIndex": 2,
      "start": 0,
      "symbol": "command",
      "type": "symbol",
    },
  ],
  "end": 61,
  "firstColumn": 1,
  "firstLine": 1,
  "id": 2,
  "internalRuleIndex": 2,
  "lastColumn": 62,
  "lastLine": 1,
  "ruleIndex": 1,
  "start": 0,
  "symbol": "progam",
  "type": "symbol",
}
`;

exports[`bash parser parse new line 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "end": 4,
              "firstColumn": 1,
              "firstLine": 1,
              "lastColumn": 5,
              "lastLine": 1,
              "start": 0,
              "text": "curl",
              "token": "WORD",
              "type": "token",
            },
          ],
          "end": 4,
          "firstColumn": 1,
          "firstLine": 1,
          "id": 4,
          "internalRuleIndex": 5,
          "lastColumn": 5,
          "lastLine": 1,
          "ruleIndex": 3,
          "start": 0,
          "symbol": "literal",
          "type": "symbol",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "end": 6,
                  "firstColumn": 6,
                  "firstLine": 1,
                  "lastColumn": 7,
                  "lastLine": 1,
                  "start": 5,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
                Object {
                  "end": 10,
                  "firstColumn": 7,
                  "firstLine": 1,
                  "lastColumn": 2,
                  "lastLine": 2,
                  "start": 6,
                  "text": "x\\\\
z",
                  "token": "STRING_CONTENT",
                  "type": "token",
                },
                Object {
                  "end": 11,
                  "firstColumn": 2,
                  "firstLine": 2,
                  "lastColumn": 3,
                  "lastLine": 2,
                  "start": 10,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
              ],
              "end": 11,
              "firstColumn": 6,
              "firstLine": 1,
              "id": 6,
              "internalRuleIndex": 10,
              "lastColumn": 3,
              "lastLine": 2,
              "ruleIndex": 4,
              "start": 5,
              "symbol": "string",
              "type": "symbol",
            },
          ],
          "end": 11,
          "firstColumn": 6,
          "firstLine": 1,
          "id": 5,
          "internalRuleIndex": 6,
          "lastColumn": 3,
          "lastLine": 2,
          "ruleIndex": 3,
          "start": 5,
          "symbol": "literal",
          "type": "symbol",
        },
        Object {
          "children": Array [
            Object {
              "end": 21,
              "firstColumn": 6,
              "firstLine": 3,
              "lastColumn": 8,
              "lastLine": 3,
              "start": 19,
              "text": "-u",
              "token": "WORD",
              "type": "token",
            },
          ],
          "end": 21,
          "firstColumn": 6,
          "firstLine": 3,
          "id": 7,
          "internalRuleIndex": 5,
          "lastColumn": 8,
          "lastLine": 3,
          "ruleIndex": 3,
          "start": 19,
          "symbol": "literal",
          "type": "symbol",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "end": 23,
                  "firstColumn": 9,
                  "firstLine": 3,
                  "lastColumn": 10,
                  "lastLine": 3,
                  "start": 22,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
                Object {
                  "end": 26,
                  "firstColumn": 10,
                  "firstLine": 3,
                  "lastColumn": 13,
                  "lastLine": 3,
                  "start": 23,
                  "text": "a:b",
                  "token": "STRING_CONTENT",
                  "type": "token",
                },
                Object {
                  "end": 27,
                  "firstColumn": 13,
                  "firstLine": 3,
                  "lastColumn": 14,
                  "lastLine": 3,
                  "start": 26,
                  "text": "\\"",
                  "token": "QUOTE",
                  "type": "token",
                },
              ],
              "end": 27,
              "firstColumn": 9,
              "firstLine": 3,
              "id": 9,
              "internalRuleIndex": 10,
              "lastColumn": 14,
              "lastLine": 3,
              "ruleIndex": 4,
              "start": 22,
              "symbol": "string",
              "type": "symbol",
            },
          ],
          "end": 27,
          "firstColumn": 9,
          "firstLine": 3,
          "id": 8,
          "internalRuleIndex": 6,
          "lastColumn": 14,
          "lastLine": 3,
          "ruleIndex": 3,
          "start": 22,
          "symbol": "literal",
          "type": "symbol",
        },
      ],
      "end": 27,
      "firstColumn": 1,
      "firstLine": 1,
      "id": 3,
      "internalRuleIndex": 4,
      "lastColumn": 14,
      "lastLine": 3,
      "ruleIndex": 2,
      "start": 0,
      "symbol": "command",
      "type": "symbol",
    },
  ],
  "end": 27,
  "firstColumn": 1,
  "firstLine": 1,
  "id": 2,
  "internalRuleIndex": 2,
  "lastColumn": 14,
  "lastLine": 3,
  "ruleIndex": 1,
  "start": 0,
  "symbol": "progam",
  "type": "symbol",
}
`;

exports[`bash parser token correctly 1`] = `
Array [
  Object {
    "channel": undefined,
    "end": 4,
    "firstColumn": 1,
    "firstLine": 1,
    "lastColumn": 5,
    "lastLine": 1,
    "start": 0,
    "t": "WORD",
    "text": "curl",
    "token": "WORD",
  },
  Object {
    "channel": "HIDDEN",
    "end": 5,
    "firstColumn": 5,
    "firstLine": 1,
    "lastColumn": 6,
    "lastLine": 1,
    "start": 4,
    "t": "HIDDEN",
    "text": " ",
    "token": "HIDDEN",
  },
  Object {
    "channel": undefined,
    "end": 6,
    "firstColumn": 6,
    "firstLine": 1,
    "lastColumn": 7,
    "lastLine": 1,
    "start": 5,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "channel": undefined,
    "end": 27,
    "firstColumn": 7,
    "firstLine": 1,
    "lastColumn": 28,
    "lastLine": 1,
    "start": 6,
    "t": "STRING_CONTENT",
    "text": "https://api.test.com/",
    "token": "STRING_CONTENT",
  },
  Object {
    "channel": undefined,
    "end": 28,
    "firstColumn": 28,
    "firstLine": 1,
    "lastColumn": 29,
    "lastLine": 1,
    "start": 27,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "channel": "HIDDEN",
    "end": 29,
    "firstColumn": 29,
    "firstLine": 1,
    "lastColumn": 30,
    "lastLine": 1,
    "start": 28,
    "t": "HIDDEN",
    "text": " ",
    "token": "HIDDEN",
  },
  Object {
    "channel": undefined,
    "end": 31,
    "firstColumn": 30,
    "firstLine": 1,
    "lastColumn": 32,
    "lastLine": 1,
    "start": 29,
    "t": "WORD",
    "text": "-u",
    "token": "WORD",
  },
  Object {
    "channel": "HIDDEN",
    "end": 32,
    "firstColumn": 32,
    "firstLine": 1,
    "lastColumn": 33,
    "lastLine": 1,
    "start": 31,
    "t": "HIDDEN",
    "text": " ",
    "token": "HIDDEN",
  },
  Object {
    "channel": undefined,
    "end": 33,
    "firstColumn": 33,
    "firstLine": 1,
    "lastColumn": 34,
    "lastLine": 1,
    "start": 32,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "channel": undefined,
    "end": 60,
    "firstColumn": 34,
    "firstLine": 1,
    "lastColumn": 61,
    "lastLine": 1,
    "start": 33,
    "t": "STRING_CONTENT",
    "text": "some_username:some_password",
    "token": "STRING_CONTENT",
  },
  Object {
    "channel": undefined,
    "end": 61,
    "firstColumn": 61,
    "firstLine": 1,
    "lastColumn": 62,
    "lastLine": 1,
    "start": 60,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "end": 61,
    "firstColumn": 62,
    "firstLine": 1,
    "lastColumn": 62,
    "lastLine": 1,
    "start": 61,
    "t": "$EOF",
    "text": "",
    "token": "$EOF",
  },
]
`;

exports[`bash parser token new line 1`] = `
Array [
  Object {
    "channel": undefined,
    "end": 4,
    "firstColumn": 1,
    "firstLine": 1,
    "lastColumn": 5,
    "lastLine": 1,
    "start": 0,
    "t": "WORD",
    "text": "curl",
    "token": "WORD",
  },
  Object {
    "channel": "HIDDEN",
    "end": 5,
    "firstColumn": 5,
    "firstLine": 1,
    "lastColumn": 6,
    "lastLine": 1,
    "start": 4,
    "t": "HIDDEN",
    "text": " ",
    "token": "HIDDEN",
  },
  Object {
    "channel": undefined,
    "end": 6,
    "firstColumn": 6,
    "firstLine": 1,
    "lastColumn": 7,
    "lastLine": 1,
    "start": 5,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "channel": undefined,
    "end": 14,
    "firstColumn": 7,
    "firstLine": 1,
    "lastColumn": 6,
    "lastLine": 2,
    "start": 6,
    "t": "STRING_CONTENT",
    "text": "x\\\\
    z",
    "token": "STRING_CONTENT",
  },
  Object {
    "channel": undefined,
    "end": 15,
    "firstColumn": 6,
    "firstLine": 2,
    "lastColumn": 7,
    "lastLine": 2,
    "start": 14,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "channel": "HIDDEN",
    "end": 16,
    "firstColumn": 7,
    "firstLine": 2,
    "lastColumn": 8,
    "lastLine": 2,
    "start": 15,
    "t": "HIDDEN",
    "text": " ",
    "token": "HIDDEN",
  },
  Object {
    "channel": "HIDDEN",
    "end": 18,
    "firstColumn": 8,
    "firstLine": 2,
    "lastColumn": 1,
    "lastLine": 3,
    "start": 16,
    "t": "HIDDEN",
    "text": "\\\\
",
    "token": "HIDDEN",
  },
  Object {
    "channel": "HIDDEN",
    "end": 23,
    "firstColumn": 1,
    "firstLine": 3,
    "lastColumn": 6,
    "lastLine": 3,
    "start": 18,
    "t": "HIDDEN",
    "text": "     ",
    "token": "HIDDEN",
  },
  Object {
    "channel": undefined,
    "end": 25,
    "firstColumn": 6,
    "firstLine": 3,
    "lastColumn": 8,
    "lastLine": 3,
    "start": 23,
    "t": "WORD",
    "text": "-u",
    "token": "WORD",
  },
  Object {
    "channel": "HIDDEN",
    "end": 26,
    "firstColumn": 8,
    "firstLine": 3,
    "lastColumn": 9,
    "lastLine": 3,
    "start": 25,
    "t": "HIDDEN",
    "text": " ",
    "token": "HIDDEN",
  },
  Object {
    "channel": undefined,
    "end": 27,
    "firstColumn": 9,
    "firstLine": 3,
    "lastColumn": 10,
    "lastLine": 3,
    "start": 26,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "channel": undefined,
    "end": 30,
    "firstColumn": 10,
    "firstLine": 3,
    "lastColumn": 13,
    "lastLine": 3,
    "start": 27,
    "t": "STRING_CONTENT",
    "text": "a:b",
    "token": "STRING_CONTENT",
  },
  Object {
    "channel": undefined,
    "end": 31,
    "firstColumn": 13,
    "firstLine": 3,
    "lastColumn": 14,
    "lastLine": 3,
    "start": 30,
    "t": "QUOTE",
    "text": "\\"",
    "token": "QUOTE",
  },
  Object {
    "end": 31,
    "firstColumn": 14,
    "firstLine": 3,
    "lastColumn": 14,
    "lastLine": 3,
    "start": 31,
    "t": "$EOF",
    "text": "",
    "token": "$EOF",
  },
]
`;
